### PR TITLE
[core][experimental] Correct `num_input_consumers` for CachedChannel

### PR DESCRIPTION
**Pull request by kevin85421**
### Problem statement

Without this PR, the `num_input_consumers` would be 1 because both `inp[0]` and `inp[1]` are only referred to in one task on the actor, so `CachedChannel` will not be created. The read will eventually time out because the mutable object is being read by the same actor twice.

```python
def test_actor_method_bind_diff_input_attr_1(ray_start_regular):
    actor = Actor.remote(0)
    with InputNode() as inp:
        # Two class methods are bound to two different input
        # attribute nodes.
        output1 = actor.inc.bind(inp[0])
        output2 = actor.inc.bind(inp[1])
        dag = MultiOutputNode([output1, output2])
    compiled_dag = dag.experimental_compile()
    ref = compiled_dag.execute(0, 1)
    assert ray.get(ref) == [0, 1]

    ref = compiled_dag.execute(1, 2)
    assert ray.get(ref) == [2, 4]

    compiled_dag.teardown()
```
image

### Solution

If a task is bound to either an `InputNode` or an `InputAttributeNode`, increment `num_input_consumers` by 1. Therefore, `num_input_consumers` will be 2, and a `CachedChannel` will be created.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/dag/tests/experimental/test_accelerated_dag.py [hunk 1]
```diff
@@ -418,6 +418,76 @@ def test_actor_method_bind_same_input_attr(ray_start_regular):
...
+        # Two class methods are bound to two different input
+        # attribute nodes.
+        output1 = actor.inc.bind(inp[0])
+        output2 = actor.inc.bind(inp[1])
+        dag = MultiOutputNode([output1, output2])
+    compiled_dag = dag.experimental_compile()
+    ref = compiled_dag.execute(0, 1)
+    assert ray.get(ref) == [0, 1]
+
+    ref = compiled_dag.execute(1, 2)
```
Comment (rkooo567) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:433]: can you also add a test for dict input? 

```
ref = compiled_dag.execute({"a:" 1, "b": 2})
```

Comment (rkooo567) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:433]: just one test is fine

**Review by rkooo567 — APPROVED** [review]

File: python/ray/dag/tests/experimental/test_accelerated_dag.py (hunk 1 above)
Comment (kevin85421) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:433]: Updated: https://github.com/ray-project/ray/pull/47489/commits/c3628810770a5f35874bd864728543e47c1a3c12

File: python/ray/dag/tests/experimental/test_accelerated_dag.py [hunk 2]
```diff
@@ -418,6 +418,94 @@ def test_actor_method_bind_same_input_attr(ray_start_regular):
...
+    assert ray.get(ref) == 4
+
+    compiled_dag.teardown()
+
+
+def test_actor_method_bind_diff_input_attr_4(ray_start_regular):
+    actor = Actor.remote(0)
+    with InputNode() as inp:
+        output1 = actor.inc_two.bind(inp[0], inp[1])
+        output2 = actor.inc.bind(inp[2])
```
Comment (ruisearch42) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:479]: Can you add a test that there are overlaps on input attributes between two calls?

```
        output1 = actor.inc_two.bind(inp[0], inp[1])
        output2 = actor.inc_two.bind(inp[2], inp[0])
```

File: python/ray/dag/compiled_dag_node.py [hunk 3]
```diff
@@ -1238,6 +1235,8 @@ def _get_or_compile(
                         assert len(upstream_task.output_channels) == 1
                         arg_channel = upstream_task.output_channels[0]
                         assert arg_channel is not None
+                if is_input_consumer:
+                    num_input_consumers += 1
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:1239]: I think we will need to compute how many tasks that consume the input, so there needs some dedup if the tasks are the same.

Say for task1(inp[0], inp[1]), it should only be count as 1

File: python/ray/dag/tests/experimental/test_accelerated_dag.py [hunk 4]
```diff
@@ -418,6 +418,94 @@ def test_actor_method_bind_same_input_attr(ray_start_regular):
...
+        # attribute node.
+        output1 = actor.inc.bind(inp[0])
+        output2 = actor.inc.bind(inp[0])
+        output3 = actor.inc.bind(inp[1])
+        dag = MultiOutputNode([output1, output2, output3])
+    compiled_dag = dag.experimental_compile()
+    ref = compiled_dag.execute(0, 1)
+    assert ray.get(ref) == [0, 0, 1]
+
+    ref = compiled_dag.execute(1, 2)
```
Comment (ruisearch42) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:453]: For all these tests, we should run multiple times to make sure the caching behavior is correct. Running once will likely miss catching issues.

I also recommend using `Collector.collect_two/three` when possible which is more powerful to check correctness.

**Review by ruisearch42 — COMMENTED** [review]
Thanks for the fix!

Need to change title?

File: python/ray/dag/tests/experimental/test_accelerated_dag.py (hunk 4 above)
Comment (kevin85421) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:453]: > For all these tests, we should run multiple times to make sure the caching behavior is correct. Running once will likely miss catching issues.

Do you mean calling `execute` more than once? Each test has already been executed twice.

> I also recommend using Collector.collect_two/three when possible which is more powerful to check correctness.

Good idea

File: python/ray/dag/compiled_dag_node.py [hunk 5]
```diff
@@ -1214,22 +1214,19 @@ def _get_or_compile(
             # or use InputAttributeNode, but not both.
             num_input_consumers = 0
 
-            # Step 1: populate num_channel_reads and perform some validation.
+            # Step 1: populate num_input_consumers and perform some validation.
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:1217]: Thanks for catching this. `arg_to_consumers` should be the main thing, but you can mention `num_input_consumers` as well.

File: python/ray/dag/compiled_dag_node.py (hunk 3 above)
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py:1239]: > Say for task1(inp[0], inp[1]), it should only be counted as 1

Is there any difference from the current behavior? The current logic is:

```python
for task in tasks:
    is_input_consumer = False
    for arg in task.args:
      # update `is_input_consumer`
    if is_input_consumer:
      num_input_consumers += 1
```

`num_input_consumers` at most increments by 1 for a task.

File: python/ray/dag/tests/experimental/test_accelerated_dag.py (hunk 4 above)
Comment (ruisearch42) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:453]: > Do you mean calling execute more than once? Each test has already been executed twice.

Yeah, maybe at least 3 times would be safer.

File: python/ray/dag/compiled_dag_node.py (hunk 5 above)
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py]: update: https://github.com/ray-project/ray/pull/47489/commits/503351449a3d2462946fb2f899ba2d6eee789efd

File: python/ray/dag/tests/experimental/test_accelerated_dag.py (hunk 4 above)
Comment (kevin85421) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:458]: update: https://github.com/ray-project/ray/pull/47489/commits/503351449a3d2462946fb2f899ba2d6eee789efd

File: python/ray/dag/tests/experimental/test_accelerated_dag.py (hunk 2 above)
Comment (kevin85421) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py]: updated: https://github.com/ray-project/ray/pull/47489/commits/acdca29ff3c95622bf47b59db01ab114baa91eed

**Comment by kevin85421** [comment]
> Need to change title?

How about "Correct num_input_consumers for CachedChannel"?

**Comment by rkooo567** [comment]
ready to be merged @kevin85421 ?

**Review by ruisearch42 — APPROVED** [review]
LGTM when tests pass!

**Comment by kevin85421** [comment]
> ready to be merged @kevin85421 ?

It will be ready to be merged after tests pass.

**Comment by rkooo567** [comment]
premerge failures

**Comment by kevin85421** [comment]
Retry the unrelated RLlib failures.

**Pull request merged**